test/umfpack.jl

```diff
@@ -365,8 +365,25 @@ end
         for nm in (:colptr, :m, :n, :nzval, :rowval, :status)
             @test getfield(F1, nm) == getfield(F2, nm)
         end
+        for nm in (:W, :Wi)
+            @test getfield(F1.workspace, nm) == getfield(F2.workspace, nm)
```

Comment:
msan complains about this because the workspace is not initialized memory. Does the workspace need to survive serialization or is it just scratch memory? If it's just scratch, maybe we shouldn't serialize it at all?

Reply:
It's just scratched, there's no reason to serialize it.

Reply:
I'll remove it from the serialization function here in a bit.

Reply:
but that wouldn't that break the semantic of serialization? suppose there are two UMFPACKLUs with shaped workspace.

Reply:
Can you elaborate Sobhan?

Reply:
well you can define two umfpacklus that share the workspace not sure what to elaborate. if you just reallocate it, it will not be the same object when de-serialized. 

something like 
L1 = lu(A)
L2 = UmfpackLU(L1.n, L1.m, another zero index sparse matrix, L1.workspace, l1.lock) 
this would work fine.

i don't think you can just reallocate. Also the size of the workspace is not determined by us. it's resized by lu (given the config) but it could have the wrong size. so by reallocating it, you might also accidentally resize it. My point is that there is no guarantee that not serializing wont break "working" code. a better idea might be to just check for length of workspace in that test.

Reply:
Serializing copies the workspace/lock anyway, it doesn't restore a reference to the original objects, so as long as the content doesn't matter to the semantics of UmfpackLU, reallocating it is fine.

Reply:
We should likely just do `serialize((size(lu.workspace.Wi), size(lu.workspace.W)))`, and provide the missing constructor to reallocate `UmfpackWS` from that tuple.

Reply:
> Serializing copies the workspace/lock anyway,

even if you serialize `(L1, L2)` ?

Reply:
works fine on master tho
```julia
using SparseArrays, Serialization, LinearAlgebra
a = lu(I + sprandn(10, 10, 0.1))
b = SparseArrays.UMFPACK.UmfpackLU(C_NULL, C_NULL, a.m, a.n,
    copy(a.colptr), copy(a.rowval), copy(a.nzval), 0, a.workspace, a.control, a.info, a.lock)
serialize("a.jld", (a, b))
x, y = deserialize("a.jld")
x.workspace == y.workspace
```
gives `true`

Reply:
I'm not really convinced that's an important feature for a workspace.

Regardless, changing the test to satisfy msan doesn't fix the underlying issue of serializing and deserializing uninitialized memory as far as I understand.

Reply:
> I'm not really convinced that's an important feature for a workspace.

i think not sanctifying this (regardless of usefulness), is a "wrong" in the sense that it does not respect the semantic of serialization. it's like saying that uninitialized vectors can be recreated on the fly instead (assuming there was a way for checking for that).

> underlying issue of serializing 

i'm not sure i understand the issue, checking for uninitialized memory reads is way of finding potential bugs and not a bug itself. any 64 bits is a valid int64 or float64. the problem is that it's possible that there is nan somewhere which i think i fixed a while back.

see https://github.com/JuliaSparse/SparseArrays.jl/blob/3b1ce54fe4575913140cc4f3b001d456da6091e9/test/umfpack.jl#L366

Reply:
also i messed up what i meant to use in the snippet above was
```julia
x.workspace === y.workspace
```
(still returns true)